File: backend/src/main/java/dev/minimtr/service/ServiceDayService.java
package dev.minimtr.service;

import dev.minimtr.config.ServiceSettings;
import dev.minimtr.model.vo.ServiceDayVo;
import dev.minimtr.repo.ServiceDayRepo;
import java.time.Instant;
import org.springframework.stereotype.Service;

@Service
public class ServiceDayService {
    private final ServiceDayRepo repo;
    private final ServiceSettings settings;

    public ServiceDayService(ServiceDayRepo repo, ServiceSettings settings) {
        this.repo = repo;
        this.settings = settings;
    }

    public ServiceDayVo load(Instant at) {
        var date = settings.serviceDateAt(at);
        var start = settings.startsAt(date);
        var end = start.plusSeconds(settings.serviceEndOffsetMinutes() * 60L);
        var state = repo.load(date);
        return new ServiceDayVo(date, !at.isBefore(start) && at.isBefore(end), start, end,
                new ServiceDayVo.Replay(state.replayStartsAt(), state.replayEndsAt()),
                new ServiceDayVo.Schedules(state.currentExists(), date.plusDays(1), state.nextExists()));
    }
}

File: server/src/services/service-day-service.ts
import type { Database } from '../db/client.js';
import { loadServiceDayDatabaseState } from '../db/service-day-queries.js';
import { nextServiceDate, serviceWindowAt } from '../domain/time.js';
import type { ServiceConfig, ServiceDaySnapshot } from '../domain/types.js';

export interface ServiceDayService {
    load(at: Date): Promise<ServiceDaySnapshot>;
}

function isoTimestamp(value: Date | string | null): string | null {
    return value ? (value instanceof Date ? value : new Date(value)).toISOString() : null;
}

export function createServiceDayService(db: Database, config: ServiceConfig): ServiceDayService {
    return {
        async load(at) {
            const window = serviceWindowAt(
                at,
                config.serviceDayStart,
                config.serviceEndOffsetMinutes,
            );
            const nextDate = nextServiceDate(window.serviceDate);
            const state = await loadServiceDayDatabaseState(db, window.serviceDate, nextDate);
            return {
                serviceDate: window.serviceDate,
                active: window.active,
                startsAt: window.startsAt.toISOString(),
                endsAt: window.endsAt.toISOString(),
                replay: {
                    startsAt: isoTimestamp(state.replay_starts_at),
                    endsAt: isoTimestamp(state.replay_ends_at),
                },
                schedules: {
                    current: state.current_exists,
                    nextServiceDate: nextDate,
                    next: state.next_exists,
                },
            };
        },
    };
}
